upload-api/ucan-invocation.js

```diff
@@ -57,6 +55,80 @@ export async function parseUcanInvocationRequest(request) {
   }
 }
 
+/**
+ * @param {UcanInvocationWrapper} ucanInvocation
+ * @param {string} ucanLogStreamName
+ */
+export function getKinesisRecord (ucanInvocation, ucanLogStreamName) {
+  return {
+    Data: uint8arrayFromString(
+      JSON.stringify({
+        carCid: ucanInvocation.carCid,
+        value: ucanInvocation.value,
+        ts: Date.now(),
+      })
+    ),
+    // https://docs.aws.amazon.com/streams/latest/dev/key-concepts.html
+    // A partition key is used to group data by shard within a stream.
+    // It is required, and now we are starting with one shard. We need to study best partition key
+    PartitionKey: 'key',
+    StreamName: ucanLogStreamName,
+  }
+}
+
+/**
+ * @typedef {object} UcanInvocationCtx
+ * @property {import('@web3-storage/upload-api').UcanBucket} storeBucket
+ * @property {string} basicAuth
+ * @property {string} [streamName]
+ * @property {import('@aws-sdk/client-kinesis').Kinesis} [kinesisClient]
+ */
+
+/**
+ * @param {import('aws-lambda').APIGatewayProxyEventV2} request
+ * @param {UcanInvocationCtx} ctx
+ */
+export async function processUcanInvocation(request, ctx) {
+  const token = getTokenFromRequest(request)
+  if (token !== ctx.basicAuth) {
+    throw new NoValidTokenError('invalid Authorization credentials provided')
+  }
+
+  const ucanInvocation = await parseUcanInvocationRequest(request)
```

Comment:
It looks like `parseUcanInvocationRequest` has an invalid assumption that CAR will contain a single invocation, while in practice that seems to be the case (mostly because how clients are written) all of the ucanto was written such that you could have bunch of the invocations send in one request (and I hope clients will start utilizing it sooner than later).

It would be very unfortunate if we bake single invocation per request assumption in this layer because we'll make it really hard to support multiple in the future. Furthermore while our clients do not send multiple invocations in one request there is no guarantee that third party clients won't so in which case we'll not persist them correctly.

Reply:
I think relevant code paths need to be updated to methods like `processUcanInvocations` and then deal with multiples in many cases concurrently.

Reply:
Thanks for raising this @Gozala 

I just addressed all code paths to take this into account. Ended up renaming some types and functions to better reflect distinction between Invocation and Car with multiple Invocations